internal/controller/postgrescluster/snapshots.go

```diff
@@ -133,49 +148,286 @@ func (r *Reconciler) reconcileVolumeSnapshots(ctx context.Context,
+		naming.LabelData:    naming.DataPostgres,
+	}
+
+	// If volume already exists, use existing name. Otherwise, generate a name.
+	var pvc *corev1.PersistentVolumeClaim
+	existingPVCName, err := getPGPVCName(labelMap, clusterVolumes)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	if existingPVCName != "" {
+		pvc = &corev1.PersistentVolumeClaim{ObjectMeta: metav1.ObjectMeta{
+			Namespace: cluster.GetNamespace(),
+			Name:      existingPVCName,
+		}}
+	} else {
+		pvc = &corev1.PersistentVolumeClaim{ObjectMeta: naming.ClusterDedicatedSnapshotVolume(cluster)}
+	}
+	pvc.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("PersistentVolumeClaim"))
 
-	// Find primary instance
-	primaryInstance := &Instance{}
-	for _, instance := range instances.forCluster {
-		if isPrimary, known := instance.IsPrimary(); isPrimary && known {
-			primaryInstance = instance
+	// If snapshots are disabled, delete the PVC if it exists and return early.
+	// Check the client cache first using Get.
+	if cluster.Spec.Backups.Snapshots == nil || !volumeSnapshotsFeatureEnabled {
+		key := client.ObjectKeyFromObject(pvc)
+		err := errors.WithStack(r.Client.Get(ctx, key, pvc))
+		if err == nil {
+			err = errors.WithStack(r.deleteControlled(ctx, cluster, pvc))
 		}
+		return nil, client.IgnoreNotFound(err)
+	}
+
+	// If we've got this far, snapshots are enabled so we should create/update/get
+	// the dedicated snapshot volume
+	pvc, err = r.createDedicatedSnapshotVolume(ctx, cluster, labelMap, pvc)
+	if err != nil {
+		return pvc, err
+	}
+
+	// Determine if we need to run a restore job, based on the most recent backup
+	// and an annotation on the PVC.
+
+	// Find the most recently completed backup job.
+	backupJob, err := r.getLatestCompleteBackupJob(ctx, cluster)
+	if err != nil {
+		return pvc, err
+	}
+
+	// Return early if no complete backup job is found.
+	if backupJob == nil {
+		return pvc, nil
 	}
-	// Return error if primary instance not found
-	if primaryInstance.Name == "" {
-		return nil, errors.New("Could not find primary instance. Cannot create volume snapshot.")
+
+	// If pvc has been annotated with latest backup job's completion time, then restore was
+	// successful and there's nothing left to do.
+	backupJobCompletionTime := backupJob.Status.CompletionTime.String()
+	// TODO: change this to check whether backupJobCompletionTime is more recent than pvc annotation.
+	if pvc.GetAnnotations()[naming.PGBackRestBackupJobCompletion] == backupJobCompletionTime {
+		return pvc, nil
+	}
+
+	// If we've made it here, the pvc has not been restored with latest backup.
+	// Find the dedicated snapshot volume restore job if it exists. Since we delete
+	// successful restores after we annotate the PVC and stop making restore jobs
+	// if a failed DSV restore job exists, there should only ever be one DSV restore
+	// job in existence at a time.
+	// TODO(dsessler7): Should this function throw an error or something if multiple
```

Comment:
I think our current setup prevents multiple restore jobs from being created, due to name collisions, right? But if there were multiple restore jobs, you'd want to make sure you retrieve the right one, because the `PGBackRestBackupJobCompletion` annotation will be propagated from whichever one is chosen.

Reply:
I can add some logic to `getDedicatedSnapshotVolumeRestoreJob` to always get the latest restore job in case there ever is more than one, but there really shouldn't ever be more than one because we only ever create a restore job if none are found (lines 243-246).